### PR TITLE
Integrate the DVM into the NFCT

### PR DESCRIPTION
**Pull request by chrismaree**
This PR integrates the DVM into the NFCT's `PricelessPositionManager` and `Liquidatable` Contracts. Integration occurs to the `PricelessPositionManager` who implements an `OracleInterface` which is used to call prices.

In `PricelessPositionManager`, the oracle price is only used at expiry of the contract. At maturity, the `expire` function is called which requests an oracle price. When this is available, token holders or sponsors can call `settleExpired` which redeems all possible tokens for their position. For a token holder this burns all tokens and sends back corresponding collateral. For a token sponsor, this will redeem all underlying after the NPV of outstanding debt. A caller can be both a holder and a sponsor and so will receive the sum of the total underlying to be paid.

In the `Liquidatable`, the oracle is only used to settle disputes after liquidations. A disputer can call `dispute` on a position which requests a price from the oracle. At a later point any one can call `settleDispute` which will use will use the oracle price to define if the total outstanding debt is more than the underlying collateral * the collateralization requirement.

### CODE REVIEW

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 1]
```diff
@@ -197,11 +226,114 @@ contract PricelessPositionManager is Testable {
...
+
+        // Get caller's tokens balance and calculate amount of underlying entitled to them.
+        FixedPoint.Unsigned memory tokensToRedeem = FixedPoint.Unsigned(tokenCurrency.balanceOf(msg.sender));
+        FixedPoint.Unsigned memory totalRedeemableCollateral = tokensToRedeem.div(settlementPrice);
+
+        // TODO: what happens in the case where the position is invalid but the contract has expired; i.e they were
+        // liquidated close to settlement and then contract settles before they withdraw.
+
+        // If the caller is a sponsor they are also entitled to their underlying excess collateral.
+        if (positions[msg.sender].isValid) {
```
Comment (chrismaree) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: this check will be removed in the next PR when we remove the `isValid` bool from contract state.

File: core/contracts/financial_templates/ExpiringMultiParty.sol [hunk 2]
```diff
@@ -12,21 +12,27 @@ contract ExpiringMultiParty is Liquidatable {
...
         FixedPoint.Unsigned memory _disputeBondPct,
         FixedPoint.Unsigned memory _sponsorDisputeRewardPct,
         FixedPoint.Unsigned memory _disputerDisputeRewardPct,
-        uint _liquidationLiveness
+        uint _liquidationLiveness,
+        address _finderAddress,
+        bytes32 _priceFeedIdentifier
     )
         public
         Liquidatable(
```
Comment (nicholaspai) [inline core/contracts/financial_templates/ExpiringMultiParty.sol]: What do you think about the tradeoffs of having ExpiringMultiParty? Is this contract even needed? just curious now that you've worked with the contracts

@mrice32 @ptare

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 3]
```diff
@@ -51,26 +59,41 @@ contract PricelessPositionManager is Testable {
...
+    modifier onlyPostExpiration() {
+        require(getCurrentTime() >= expirationTimestamp, "Cannot operate on a position before its expiry time");
+        _;
+    }
+
     /**
      * @notice Transfers ownership of the caller's current position to `newSponsorAddress`. The address
      * `newSponsorAddress` isn't allowed to have a position of their own before the transfer.
      */
+    // TODO: transfer should not work if there is a pending withdraw
```
Comment (nicholaspai) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: I know you've highlighted as a TODO, but why not implement it now? It's a one liner right?

`require(positionData.requestPassTimestamp == 0, "Cannot transfer with a pending withdrawal request");`

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 4]
```diff
@@ -125,6 +151,8 @@ contract PricelessPositionManager is Testable {
      * @notice Starts a withdrawal request that, if passed, allows the sponsor to withdraw `collateralAmount` from their
      * position. The request will be pending for `withdrawalLiveness`, during which the position can be liquidated.
      */
+    // TODO: should this check if the position is valid first?
+    // TODO: should this check if the contract is pre-expiration?
```
Comment (nicholaspai) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: isn't this already captured in the require statement?

```
require(
            requestPassTime < expirationTimestamp,
            "Cannot request withdrawal that would pass after contract expires"
        );
```

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 5]
```diff
@@ -197,11 +226,114 @@ contract PricelessPositionManager is Testable {
...
+     * @notice After a contract has passed maturity all token holders can redeem their tokens for underlying at
+     * the prevailing price defined by the DVM from the `expire` function. This Burns all tokens from the caller
+     * of `tokenCurrency` and sends back the proportional amount of `collateralCurrency`.
+     */
+    function settleExpired() public onlyPostExpiration() {
+        // Get the current settlement price. If it is not resolved will revert.
+        FixedPoint.Unsigned memory settlementPrice = _getOraclePrice(expirationTimestamp);
+
+        // Get caller's tokens balance and calculate amount of underlying entitled to them.
+        FixedPoint.Unsigned memory tokensToRedeem = FixedPoint.Unsigned(tokenCurrency.balanceOf(msg.sender));
```
Comment (nicholaspai) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: So token holders must redeem all at once? Any tradeoffs to not allowing partial redemptions? FWIW, I like the simplicity

File: core/test/financial_templates/Liquidatable.js [hunk 6]
```diff
@@ -256,61 +292,63 @@ contract("Liquidatable", function(accounts) {
...
-        );
         // Mint enough tokens to disputer for another dispute bond
         await collateralToken.mint(disputer, disputeBond, { from: contractDeployer });
+        await collateralToken.increaseAllowance(liquidationContract.address, disputeBond, { from: disputer });
         assert(
           await didContractThrow(liquidationContract.dispute(liquidationParams.uuid, sponsor, { from: disputer }))
         );
         assert.equal((await collateralToken.balanceOf(disputer)).toString(), disputeBond.toString());
       });
+      // Weird edge cases, test anyways:
```
Comment (nicholaspai) [inline core/test/financial_templates/Liquidatable.js]: Why not delete these?

File: core/contracts/financial_templates/ExpiringMultiParty.sol (hunk 2 above)
Comment (chrismaree) [reply core/contracts/financial_templates/ExpiringMultiParty.sol]: I dont think we need it in its current form. However I think that @mrice32 is going to create a parent contract which will sit at the top of the three levels for fee collection.

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 3 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: It's for the next PR. I didnt want to add functionality over and above DVM integration here :)

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 4 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: you right.

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 5 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: ye, all or nothing. I thought about this one but didnt see a situation in which a token holder would want to redeem anything less than their current full balance. Open to changing it but this is defos easier

File: core/test/financial_templates/Liquidatable.js (hunk 6 above)
Comment (chrismaree) [reply core/test/financial_templates/Liquidatable.js]: I actually think some of these should be included 🤔 . Will look a bit deeper and update the PR

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 7]
```diff
@@ -197,11 +226,114 @@ contract PricelessPositionManager is Testable {
...
+        FixedPoint.Unsigned memory totalRedeemableCollateral = tokensToRedeem.div(settlementPrice);
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: Am I missing something or should this be:
```
tokensToRedeem.mul(settlementPrice);
```
??

Isn't `collateral_owed = token_price * num_tokens`?

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 8]
```diff
@@ -197,11 +226,114 @@ contract PricelessPositionManager is Testable {
...
+
+        // TODO: what happens in the case where the position is invalid but the contract has expired; i.e they were
+        // liquidated close to settlement and then contract settles before they withdraw.
+
+        // If the caller is a sponsor they are also entitled to their underlying excess collateral.
+        if (positions[msg.sender].isValid) {
+            PositionData storage positionData = _getPositionData(msg.sender);
+
+            // Calculate the underlying entitled to a token sponsor. This is collateral - debt in underlying.
+            FixedPoint.Unsigned memory tokenDebtValueInCollateral = positionData.tokensOutstanding.div(settlementPrice);
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: Same comment as above.

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 9]
```diff
@@ -197,11 +226,114 @@ contract PricelessPositionManager is Testable {
...
+    function expire() public onlyPostExpiration() {
+        _requestOraclePrice(expirationTimestamp);
+    }
+
+    /**
+     * @notice After a contract has passed maturity all token holders can redeem their tokens for underlying at
+     * the prevailing price defined by the DVM from the `expire` function. This Burns all tokens from the caller
+     * of `tokenCurrency` and sends back the proportional amount of `collateralCurrency`.
+     */
+    function settleExpired() public onlyPostExpiration() {
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: Does this method cover the case where there's not enough collateral in the contract to cover all the token holders?

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 10]
```diff
@@ -197,11 +226,114 @@ contract PricelessPositionManager is Testable {
...
+            FixedPoint.Unsigned memory tokenDebtValueInCollateral = positionData.tokensOutstanding.div(settlementPrice);
+            FixedPoint.Unsigned memory positionRedeemableCollateral = positionData.collateral.sub(
+                tokenDebtValueInCollateral
+            );
+
+            // Add the number of redeemable tokens for the sponsor to their total redeemable collateral.
+            totalRedeemableCollateral = totalRedeemableCollateral.add(positionRedeemableCollateral);
+
+            // Reset the position state as all the value has been removed after settlement.
+            positionData.collateral = FixedPoint.fromUnscaledUint(0);
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: nit: instead of manually setting to `0`, can we just delete the positionData struct in its entirety?

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 11]
```diff
@@ -197,11 +226,114 @@ contract PricelessPositionManager is Testable {
...
+        PositionData storage positionToLiquidate = _getPositionData(sponsor);
+
+        // Remove the collateral and outstanding from the overall total position.
+        totalPositionCollateral = totalPositionCollateral.sub(positionToLiquidate.collateral);
+        totalTokensOutstanding = totalTokensOutstanding.sub(positionToLiquidate.tokensOutstanding);
+
+        // Reset the sponsors position to have zero outstanding and collateral.
+        positionToLiquidate.tokensOutstanding = FixedPoint.fromUnscaledUint(0);
+        positionToLiquidate.collateral = FixedPoint.fromUnscaledUint(0);
+        positionToLiquidate.isValid = false;
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: nit: similar to above, can we just delete the positionData?

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 9 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: not as of yet.

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 7 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: you are 100% correct. I have fixed this and pushed updated code and tests.

File: core/contracts/financial_templates/ExpiringMultiParty.sol (hunk 2 above)
Comment (mrice32) [reply core/contracts/financial_templates/ExpiringMultiParty.sol]: I think this is something we can worry about once we have the contract working since it's more of a superficial thing.

File: core/contracts/financial_templates/Liquidatable.sol [hunk 12]
```diff
@@ -240,35 +249,30 @@ contract Liquidatable is PricelessPositionManager {
...
+
+        // The required collateral is the value of the tokens in underlying * required collateral ratio.
+        FixedPoint.Unsigned memory requiredCollateral = tokenRedemptionValue.mul(collateralRequirement);
 
-        // TODO: For testing purposes
-        disputedLiquidation.settlementPrice = hardcodedPrice;
+        // If the position has more than the required collateral it is solvent and the dispute is valid(liquidation is invalid)
+        // Note that this check uses the liquidatedCollateral not the lockedCollateral as this considers withdrawals.
+        // bool disputeSucceeded = !requiredCollateral.isLessThan(disputedLiquidation.liquidatedCollateral);
+        bool disputeSucceeded = disputedLiquidation.liquidatedCollateral.isGreaterThan(requiredCollateral);
```
Comment (mrice32) [inline core/contracts/financial_templates/Liquidatable.sol]: nit: this should be greater than or equal to.

File: core/contracts/financial_templates/Liquidatable.sol [hunk 13]
```diff
@@ -320,12 +316,7 @@ contract Liquidatable is PricelessPositionManager {
             } else if (msg.sender == sponsor) {
                 // Pay SPONSOR: remaining collateral (locked collateral - TRV) + sponsor reward
                 FixedPoint.Unsigned memory remainingCollateral;
-                // if TRV > Collateral, there is 0 excess collateral
-                if (tokenRedemptionValue.isGreaterThan(liquidation.lockedCollateral)) {
-                    remainingCollateral = FixedPoint.fromUnscaledUint(0);
-                } else {
-                    remainingCollateral = liquidation.lockedCollateral.sub(tokenRedemptionValue);
-                }
+                remainingCollateral = liquidation.lockedCollateral.sub(tokenRedemptionValue);
```
Comment (mrice32) [inline core/contracts/financial_templates/Liquidatable.sol]: Two general comments about the way this is structured: what prevents some party (like the sponsor) from calling multiple times and getting their reward twice?

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 14]
```diff
@@ -197,11 +226,111 @@ contract PricelessPositionManager is Testable {
...
+    function _getOracleAddress() internal view returns (address) {
+        bytes32 oracleInterface = "Oracle";
+        return finder.getImplementationAddress(oracleInterface);
+    }
+
+    function _getIdentifierWhitelistAddress() internal view returns (address) {
+        bytes32 identifierWhitelistInterface = "IdentifierWhitelist";
+        return finder.getImplementationAddress(identifierWhitelistInterface);
+    }
+    function _getStoreAddress() internal view returns (address) {
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: nit: is this used anywhere?

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 15]
```diff
@@ -197,11 +226,111 @@ contract PricelessPositionManager is Testable {
...
+        }
+        return FixedPoint.Unsigned(_safeUintCast(oraclePrice));
+    }
+
+    function _getOracleAddress() internal view returns (address) {
+        bytes32 oracleInterface = "Oracle";
+        return finder.getImplementationAddress(oracleInterface);
+    }
+
+    function _getIdentifierWhitelistAddress() internal view returns (address) {
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: nit: is this used anywhere?

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 16]
```diff
@@ -197,11 +226,111 @@ contract PricelessPositionManager is Testable {
...
+        int oraclePrice = oracle.getPrice(priceIdentifer, requestedTime);
+
+        // For now we don't want to deal with negative prices in positions.
+        if (oraclePrice < 0) {
+            oraclePrice = 0;
+        }
+        return FixedPoint.Unsigned(_safeUintCast(oraclePrice));
+    }
+
+    function _getOracleAddress() internal view returns (address) {
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: nit: just to further simplify things for the caller -- what about doing the cast to OracleInterface in this method and just returning the type `OracleInterface`?

File: core/test/financial_templates/ExpiringMultiParty.js [hunk 17]
```diff
@@ -1,23 +1,24 @@
...
-      "1000",
-      collateralAddress.address,
-      { rawValue: toWei("0.1") },
-      { rawValue: toWei("0.1") },
-      { rawValue: toWei("0.1") },
-      1000
-    );
-  });
-});
+// contract("ExpiringMultiParty", function(accounts) {
```
Comment (mrice32) [inline core/test/financial_templates/ExpiringMultiParty.js]: Why is this commented out?

File: core/test/financial_templates/Liquidatable.js [hunk 18]
```diff
@@ -256,7 +292,18 @@ contract("Liquidatable", function(accounts) {
...
         assert.equal(liquidation.disputeTime.toString(), disputeTime.toString());
       });
-      it("Liquidation has already been disputed", async () => {
+      it("Dispute initiates an oracle call", async () => {
+        const disputeTime = await liquidationContract.getCurrentTime();
+        await liquidationContract.dispute(liquidationParams.uuid, sponsor, { from: disputer });
+        // Oracle should have an enqueued price after calling dispute
+        const pendingRequests = await mockOracle.getPendingQueries();
+        assert.equal(
+          pendingRequests[0]["identifier"],
```
Comment (mrice32) [inline core/test/financial_templates/Liquidatable.js]: I think I realized why we've never had to do this padding: we generally compare in utf8 space, not hex. So, you should be able to get around the padding by:
```
const { hexToUtf8 } = web3.utils; // put this somewhere way up top :)
assert.equal(hexToUtf8(pendingRequests[0]["identifier"]), hexToUtf8(priceTrackingIdentifier));
```

The nice consequence of this is that when you get failures, you see the human readable string and not the hex.

File: core/test/financial_templates/Liquidatable.js [hunk 19]
```diff
@@ -256,7 +292,18 @@ contract("Liquidatable", function(accounts) {
...
+          priceTrackingIdentifier.padEnd(pendingRequests[0]["identifier"].length, 0)
+        );
+        assert.equal(pendingRequests[0]["time"], disputeTime);
```
Comment (mrice32) [inline core/test/financial_templates/Liquidatable.js]: tiny style nit: prefer `.` syntax for accessing js objects:
```
pendingRequests[0].time
```

Here and elsewhere.

File: core/contracts/financial_templates/Liquidatable.sol (hunk 12 above)
Comment (chrismaree) [reply core/contracts/financial_templates/Liquidatable.sol]: it should indeed. I tried to implement this using `!isLessThan` but had issued with this. will park this for now until we implement a `isGreatThanOrEqual` in the `FixedPoint` lib.

Comment (chrismaree) [reply core/contracts/financial_templates/Liquidatable.sol]: I will try again with the `!isLessThan` notation and see how it goes.

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 14 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: not yet...Can remove it :)

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 15 above)
Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: same as above

File: core/test/financial_templates/ExpiringMultiParty.js (hunk 17 above)
Comment (chrismaree) [reply core/test/financial_templates/ExpiringMultiParty.js]: right now we dont need it in any way. it's just a placeholder. I can remove it or implement the interface fully. open for input here

File: core/test/financial_templates/Liquidatable.js (hunk 18 above)
Comment (chrismaree) [reply core/test/financial_templates/Liquidatable.js]: ok great! will update the test :)

File: core/test/financial_templates/Liquidatable.js (hunk 19 above)
Comment (chrismaree) [reply core/test/financial_templates/Liquidatable.js]: noted, will update

File: core/contracts/financial_templates/Liquidatable.sol (hunk 13 above)
Comment (chrismaree) [reply core/contracts/financial_templates/Liquidatable.sol]: will think about this and get back to you.

File: core/test/financial_templates/ExpiringMultiParty.js (hunk 17 above)
Comment (mrice32) [reply core/test/financial_templates/ExpiringMultiParty.js]: IMO, as long as the ExpiringMultiparty class exists and has a constructor, we should have at least one test that runs that constructor to make sure it's not broken.

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 14 above)
Comment (mrice32) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: I would remove because we don't know if either of these are going to be used in this class. For instance, I think the store may only be used in the FeePayer class (depending on how we implement it).

Comment (mrice32) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: So it may have the effect of just increasing code churn.

Comment (chrismaree) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: cool. will ditch

File: core/test/financial_templates/ExpiringMultiParty.js (hunk 17 above)
Comment (chrismaree) [reply core/test/financial_templates/ExpiringMultiParty.js]: ok will do.

File: core/test/financial_templates/PricelessPositionManager.js [hunk 20]
```diff
@@ -311,6 +345,93 @@ contract("PricelessPositionManager", function(accounts) {
...
+    // Settling an expired position should revert if the contract has expired but the DVM has not yet returned a price.
+    assert(await didContractThrow(pricelessPositionManager.settleExpired({ from: tokenHolder })));
+
+    // Push a settlement price into the mock oracle to simulate a DVM vote. Say settlement occurs at 1500 usd per eth.
+    // This will result in the large position created in the beginning being at 150% collateralization with its
+    // 1000 tokens of debt, valued at $1 per token, and 1 token of collateral, valued at $1500 per token.
+    const redemptionPrice = toWei("0.0005");
+    await mockOracle.pushPrice(priceTrackingIdentifier, expirationTime.toNumber(), redemptionPrice);
+
+    //Token holder settle their synthetic tokens and redeem their tokens.
```
Comment (mrice32) [inline core/test/financial_templates/PricelessPositionManager.js]: nit: comment style

File: core/test/financial_templates/PricelessPositionManager.js [hunk 21]
```diff
@@ -311,6 +345,93 @@ contract("PricelessPositionManager", function(accounts) {
     assert(await didContractThrow(pricelessPositionManager.transfer(accounts[3], { from: sponsor })));
   });
 
+  it("Redemption post expiry", async function() {
```
Comment (mrice32) [inline core/test/financial_templates/PricelessPositionManager.js]: Just FYI: I think this is a really solid test: very clean, well commented, and understandable 💯

File: core/test/financial_templates/PricelessPositionManager.js [hunk 22]
```diff
@@ -311,6 +345,93 @@ contract("PricelessPositionManager", function(accounts) {
...
+    // The value in underlying is the number of tokens they held in the beginning/settlement price
+    assert(
+      tokenHolderFinalCollateral.sub(tokenHolderInitialCollateral),
+      tokenHolderInitialCollateral.div(toBN(redemptionPrice))
+    );
+
+    // The token holder should have no synthetic positions left after settlement.
+    assert(tokenHolderFinalSynthetic, 0);
+
+    //Token sponsor settle their synthetic tokens and redeem their tokens.
```
Comment (mrice32) [inline core/test/financial_templates/PricelessPositionManager.js]: nit: comment style.

File: core/test/financial_templates/PricelessPositionManager.js [hunk 23]
```diff
@@ -140,6 +173,8 @@ contract("PricelessPositionManager", function(accounts) {
     sponsorFinalBalance = await collateral.balanceOf(sponsor);
     assert.equal(sponsorFinalBalance.sub(sponsorInitialBalance).toString(), expectedSponsorCollateral);
     await checkBalances(toBN("0"), toBN("0"));
+
+    // TODO: Add a test to check that normal redemption does not work after expiratory.
```
Comment (mrice32) [inline core/test/financial_templates/PricelessPositionManager.js]: nit: typo `expiratory`.

File: core/contracts/financial_templates/Liquidatable.sol (hunk 13 above)
Comment (chrismaree) [reply core/contracts/financial_templates/Liquidatable.sol]: There is one point where this could be a problem. I have added a todo.

Comment (mrice32) [reply core/contracts/financial_templates/Liquidatable.sol]: Sweet, thanks!
